---
title: "WELL_China_Microbiome_Dada2"
author: "Yan Min"
date: "9/6/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This file serves as a documentation of WELL-China microbiome study. Firstly, import raw sequencing data into R and analyze into a phyloseq object including a taxa table, phylogenetic tree, otu table, sample data, and reference seq. 

Install and load DADA2.

```{r packages}
source("http://bioconductor.org/biocLite.R")
biocLite("dada2")
biocLite("shiny")
graphPkgs <- c("ggplot2", "gridExtra")
install.packages(graphPkgs)
library(dada2)
library(ggplot2)
library(gridExtra)
library(shiny)
```

Load packages and set up working directory.

```{r pressure, echo=FALSE}
setwd("~/Documents/WELL_Project/WELL-China_Microbiome/3rdRoundAnalysis")
path <- "~/Documents/WELL_Project/WELL-China_Microbiome/3rdRoundAnalysis"
```

Sort forward and reverse reads, make sure they are in the same order, using the same naming convention.

```{r sort forward and reverse reads}
fnFs <- sort(list.files(path, pattern = "_R1_001.fastq"))
fnRs <- sort(list.files(path, pattern = "_R2_001.fastq"))
```

Extract sample names, and specify the full path to the fnFs and fnRs

```{r sample names}
sample.names <- sapply(strsplit(fnFs, "_"), `[`, 1)
fnFs <- file.path(path, fnFs)
fnRs <- file.path(path, fnRs)
```

Plot out quality profiles of forward and reverse reads.

```{r quality}
QualF <- plotQualityProfile(fnFs[1:2])
QualR <- plotQualityProfile(fnRs[1:2]) 
```

Filtering and trimming the raw data based on the quality profile. 

```{r filter}
filt_path <- file.path(path, "filtered")
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen = c(150, 150), 
                     maxN = 0, maxEE = c(2,2), truncQ = 2, rm.phix = TRUE, 
                     compress = TRUE, multithread = TRUE)
head(out)
```

Inspect the error rates in both forward and reverse readings.

```{r error rates}
errF <- learnErrors(filtFs, multithread = TRUE)
errR <- learnErrors(filtRs, multithread = TRUE)
plotErrF <- plotErrors(errF, nominalQ = TRUE) 
plotErrR <- plotErrors(errR, nominalQ = TRUE)
grid.arrange(nrow = 1, plotErrF, plotErrR)
```

Dereplication - combines all the identical sequencing reads into "unique sequences" with a corresponding "abundance": the number of reads with that unique sequence.

```{r dereplication}
derepFs <- derepFastq(filtFs, verbose = TRUE)
derepRs <- derepFastq(filtRs, verbose = TRUE)
names(derepFs) <- sample.names
names(derepRs) <- sample.names
```

Apply core sequence-variant inference algorithm to the dereplicated data using 
divisive amplicon denoising algorithm (DADA)

```{r dada}
dadaFs <- dada(derepFs, err = errF, multithread = TRUE)
dadaRs <- dada(derepRs, err = errR, multithread = TRUE)
```

Merge the denoised forward and reverse reads. 
```{r mergers}
mergers <- mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose = TRUE)
```

Construct a sequence table of these 221 samples. 
```{r seq table}
seqtab <- makeSequenceTable(mergers)
```

Remove chimeras
```{r chimeras}
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread = TRUE, verbose = TRUE)
sum(seqtab.nochim)/sum(seqtab) # The chimera takes up to 1% of the total sequence reads. 
```

Track reads through the pipeline.
```{r track}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track) <- sample.names
head(track)
```

Assign taxonomy using the GreenGenes and Silva reference database. 
```{r}
taxa <- assignTaxonomy(seqtab.nochim, "~/Documents/WELL_Project/WELL-China_Microbiome/3rdRoundAnalysis/rdp_train_set_14.fa.gz", multithread = TRUE)
colnames(taxa) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")
unname(head(taxa))
```

Construct the phyloseq object. 

Firstly download required pacakges.
```{r phyloseq}
source("http://bioconductor.org/biocLite.R")
biocLite("Phyloseq")
biocLite("DECIPHER")
biocLite("phangorn")
```

Load the phyloseq package to console.
```{r load package}
library(phyloseq)
library(DECIPHER)
library(phangorn)
library(dplyr)
```

Construct a phylogenetic tree using DEFIPHER and phangorn.
Firstly, construct a neighbor-joining tree, and then fit a GTR+G+I (Generalized time-reversible with Gamma rate variation) maximum likelihood tree using the neighbor-joining tree as a starting point.
```{r tree}
seqs <- getSequences(seqtab.nochim)
names(seqs) <- seqs # this propogates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor = NA)
phang.align <- phyDat(as(alignment, "matrix"), type = "DNA")
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) # tip order != sequence order
fit <- pml(treeNJ, data = phang.align)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE, rearrangement = "stochastic", control = pml.control(trace = 0))
```

Combine sequencing data with the sample data.
```{r construct phyloseq object}
samdf <- read.csv(file = "WELL_China_Microbiome_DXA_Data_Uncleaned.csv", header = TRUE, sep = ",")
samdf <- samdf[,-1]
samples.out <- rownames(seqtab.nochim)
row.names(samdf) <- samples.out
row.names(scaleDxa) <- samples.out
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), sample_data(dxaSub), tax_table(taxa), phy_tree(fitGTR$tree))
ps_scaleDxaAll <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), sample_data(scaleDxa), tax_table(taxa))
save.image("~/Documents/WELL_Project/WELL-China_Microbiome/3rdRoundAnalysis/phyloseq.rdata")
```